### PR TITLE
build cryptography=3.4.8 for python 3.10.*

### PR DESCRIPTION
**Pull request by fhoehle**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] ~Bumped~ Changed the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Some packages like `snowflake-connector-python` restricting updates of `cryptography` by excluding the next major version i.e. `<4`. In order to progress with the `python=3.10` migration it would be helpful to build latest `cryptography=3.*` also for `python=3.10`. This PR could start some `abi_branch` if there are no objects. Please note: this PR **must not** be merged to master but to some different branch `3.4.x` maybe.

### CODE REVIEW

**Comment by fhoehle** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
@conda-forge/cryptography 
Do we want to do this?
If we do, I could create a branch for this, but given the infra changes since then (cross-compilation & various fixes), I'd actually suggest to branch off from current master rather than from before #71 was merged.

**Comment by jakirkham** [comment]
@fhoehle can you please share why this is needed? This is quite an old version of a package used for...well...cryptography

**Comment by h-vetinari** [comment]
> 
> 
> @fhoehle can you please share why this is needed? This is quite an old version of a package used for...well...cryptography

In his defense, 3.4.8 is [barely](https://github.com/pyca/cryptography/releases/tag/3.4.8) 3 months old (only that upstream changed their version scheme since then), and it's also the last version that doesn't hard-require rust support to build.

**Comment by fhoehle** [comment]
> @fhoehle can you please share why this is needed? This is quite an old version of a package used for...well...cryptography

as @h-vetinari said it is not very old it is the second latest release from August 24 2021. Reasoning for is that a transition to `python=3.10` does not require a simultaneous transition from `cryptography=3.4.*` to `cryptography=35.*`.

@h-vetinari 
> If we do, I could create a branch for this, but given the infra changes since then (cross-compilation & various fixes), I'd actually suggest to branch off from current master rather than from before #71 was merged.

I branched off at master, see parent commit of this branch:
grafik

**Comment by h-vetinari** [comment]
OK, I created a branch and pointed your PR to it. There's now a conflict to solve, and you'll have to remove the unconditional `skip: true` that I added so that the CI doesn't generate anything for the new branch yet.

The build number is already correct, there was only one [build](https://anaconda.org/conda-forge/cryptography/files?page=2) (`_0`) for 3.4.8.

**Comment by fhoehle** [comment]
> OK, I created a branch and pointed your PR to it. There's now a conflict to solve, and you'll have to remove the unconditional `skip: true` that I added so that the CI doesn't generate anything for the new branch yet.

thanks @h-vetinari. I rebased on `3.4.x` and adapted the cross compiling patch to be applicable for `cryptography=3.4.8` updating the shipped Rust dependency `pyo3` from `0.13.1` to `0.15.0`. This fixes the Python 3.10 building problems on `osx-arm64`.

File: recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch
Comment (h-vetinari) [inline recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch:3]: Please keep the attribution in the patch as it was.

**Comment by h-vetinari** [comment]
I think this PR is fine to merge and poses no risk, but CC @conda-forge/core for visibility since we're going beyond what upstream does on a security-relevant package.

**Comment by beckermr** [comment]
How did these two migrations become coupled?

File: recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch
Comment (beckermr) [reply recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch:3]: I second this.

**Comment by h-vetinari** [comment]
> How did these two migrations become coupled?

`cryptography` doesn't have a migration. It's just that the 3.10 migrator came by after this feedstock had already been bumped to 3.4.8.

At first I thought that upstream only started supporting python 3.10 with v35.0.0 as well, but actually, they've (only) been [releasing](https://pypi.org/project/cryptography/3.4.8/#files) `abi3` wheels, so their 3.4.8 wheels can still be installed on 3.10 (just tested this).

File: recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch
Comment (fhoehle) [reply recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch]: Since the patch of `cryptography=35.0.0` was not applicable I generate one based on `cryptography=3.4.8` replacing the existing patch. But fine with me, to change the header, I do not think it is relevant for patching.

Comment (h-vetinari) [reply recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch]: Rebasing is quite a different thing from coming up with it in the first place (which often contains lots of tries that end up not getting reflected in the final commit); so keeping that information is mostly about who to ask if something doesn't work anymore (and not having to go hunt for it in the git history).

**Review by h-vetinari — APPROVED** [review]
I think we should do this, but I'd like to have another approval from @conda-forge/cryptography or core.

To be clear: I was wrong in my original ping above: upstream supports 3.4.8 for python 3.10 by virtue of `abi3` wheels.

**Comment by beckermr** [comment]
Cc @conda-forge/core any thoughts here?

**Comment by xhochy** [comment]
👍 on this PR, I expect that we will see many more of these backports PRs for the 3.10 migration (like we did for past migrations).

**Pull request merged**
